File: third semester/task1_2/task2/task2/Simulation.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace task2
{
    /// <summary>
    /// Class for simulation network.
    /// </summary>
    public class Simulation
    {
        private StorageNetworkTopology networkTopology = new StorageNetworkTopology();
        private Random rand;

        public Simulation(Random currentRand)
        {
            rand = currentRand;
            networkTopology.TryInfect(100, 0);
        }

        /// <summary>
        /// Change states of compurets. Try to infect them.
        /// </summary>
        public void ChangeNetworkState()
        {
            bool[] infected = new bool[networkTopology.SizeOfTopology()];
            for (int i = 0; i < networkTopology.SizeOfTopology(); ++i)
            {
                infected[i] = networkTopology.IsInfected(i);
            }

            int temp = rand.Next(1, 100);
            for (int i = 0; i < networkTopology.SizeOfTopology(); ++i)
            {
                if (infected[i])
                {
                    for (int j = 0; j < networkTopology.SizeOfTopology(); ++j)
                    {
                        if (networkTopology.Relations(i, j))
                        {
                            networkTopology.TryInfect(temp, j);
                        }
                    }
                }
            }
        }

        /// <summary>
        /// Show state of current computer.
        /// </summary>
        /// <param name="currentComputer">Number of computer.</param>
        /// <returns>State of computer.</returns>
        public string ShowState(int currentComputer)
        {
            if (networkTopology.IsInfected(currentComputer))
            {
                return "is infected.";
            }
            else
            {
                return "is healthy.";
            }
        }

        /// <summary>
        /// Get value about size of network.
        /// </summary>
        /// <returns>Size of this network.</returns>
        public int NumberOfComputers()
        {
            return networkTopology.SizeOfTopology();
        }

        /// <summary>
        /// Show states of all computers in network.
        /// </summary>
        public void ShowAllStates()
        {
            for (int i = 0; i < networkTopology.SizeOfTopology(); ++i)
            {
                System.Console.WriteLine(i.ToString() + ' ' + ShowState(i));   
            }
        }
    }
}

File: fourth semester/task5_1/task5_1/Simulation.fs
﻿module Simulation 
open StorageNetworkTopology
open System

   type Simulation(random : System.Random, input : List<int * string>, matrix : int [][], needSize : int) = 
      let mutable networkTopology : StorageNetworkTopology = StorageNetworkTopology(input, matrix, needSize)
      let mutable infected : bool array = Array.zeroCreate needSize
      let mutable array : int array = Array.create needSize -1

      member v.ChangeNetworkState = 
         networkTopology.ProcessingDataOfNetwork
         let rec getInfectedArray count =
             if (count < networkTopology.SizeOfTopology()) then
                infected.[count] <- networkTopology.IsInfected(count)
                let count = count + 1
                getInfectedArray count
         getInfectedArray 0
         let randNext = random.Next()
         let mutable count : int = 0

         for i in 0 .. needSize - 1 do
            if (infected).[i] = true then
               for j in 0 .. needSize - 1 do
               if networkTopology.Relations i j then
                  array.[count] <- j
                  count <- count + 1

         let length = Array.length array
         let rec go i =
            if (i < length) then
               if (array).[i] <> -1 then
                  networkTopology.TryInfect randNext (array).[i] 
               let i = i + 1
               go i
         go 0 

      member v.ShowState (currentComputer : int) =
         if (networkTopology.IsInfected(currentComputer)) then
            "is infected."
         else
            "is healthy."

      member v.ShowAllStates =
        let rec show i = 
            if i < networkTopology.SizeOfTopology() then
               let number = string i;
               let separate = string " "
               printfn "%A" (number + separate + v.ShowState(i))
               let i = i + 1
               show i
        show 0  